src/io.c

```diff
@@ -595,163 +609,105 @@ create_dir (const gchar *dir)
-
-            g_io_channel_unref (chan);
-
-            uzbl.io->socket_path = g_strdup (path);
-            uzbl_events_send (SOCKET_SET, NULL,
-                TYPE_STR, uzbl.io->socket_path,
-                NULL);
-            /* TODO: Collect all environment settings into one place. */
-            g_setenv ("UZBL_SOCKET", uzbl.io->socket_path, TRUE);
-            return TRUE;
-        }
-    } else {
-        g_warning ("attach_socket: could not bind to %s: %s\n", path, strerror (errno));
+    GOutputStream *os = g_io_stream_get_output_stream (G_IO_STREAM (stream));
+    if (!g_output_stream_close (os, NULL, &error)) {
+        g_warning ("Failed to close write end of fifo: %s\n", error->message);
+        g_error_free (error);
+        return FALSE;
     }
 
-    return FALSE;
+    add_buffered_cmd_source (G_IO_STREAM (stream), "Uzbl main fifo",
+                             control_command_stream, NULL, NULL);
+    uzbl.io->fifo_path = g_strdup (path);
+    uzbl_events_send (FIFO_SET, NULL,
+                      TYPE_STR, uzbl.io->fifo_path,
+                      NULL);
+    /* TODO: Collect all environment settings into one place. */
+    g_setenv ("UZBL_FIFO", uzbl.io->fifo_path, TRUE);
+    return TRUE;
 }
 
-void
-send_buffered_event (gpointer event, gpointer data)
-{
-    UZBL_UNUSED (data);
-
-    const gchar *message = (const gchar *)event;
-
-    send_event_sockets (uzbl.io->connect_sockets, message);
-}
+static void
+accept_socket_cb (GObject *source, GAsyncResult *res, gpointer data);
 
 gboolean
-line_buffer_io (GIOChannel *gio, GIOCondition condition, gpointer data)
+attach_socket (const gchar *path)
 {
-    UzblIOBufferData *io_data = (UzblIOBufferData *)data;
-
-    static const gsize BUFSZ = 256;
-    gchar input[BUFSZ];
-    gsize read;
     GError *error = NULL;
+    GSocketAddress *addr = g_unix_socket_address_new (path);
+    GSocket *sock = g_socket_new (G_SOCKET_FAMILY_UNIX,
+                                  G_SOCKET_TYPE_STREAM, 0, &error);
 
-    if ((condition & G_IO_IN) == G_IO_IN) {
-        do {
-            GIOStatus status = g_io_channel_read_chars (gio, input, BUFSZ, &read, &error);
-
-            if (status == G_IO_STATUS_ERROR) {
-                g_warning ("Error buffering: %s", error->message);
-                g_clear_error (&error);
-
-                if (io_data->error_callback) {
-                    io_data->error_callback(gio, io_data->data);
-                }
+    if (!sock) {
+        g_warning ("attach_socket: failed to create socket %s\n", error->message);
+        g_error_free (error);
+        return FALSE;
+    }
 
-                return G_SOURCE_REMOVE;
-            }
+    if (!g_socket_bind (sock, addr, FALSE, &error)) {
+        g_warning ("attach_socket: could not bind to %s: %s\n", path, error->message);
+        g_error_free (error);
+        return FALSE;
+    }
 
-            if (status == G_IO_STATUS_EOF) {
-                break;
-            }
+    if (chmod (path, 0700)) {
```

Comment:
no. that should not be necessary. This is just keeping the current way but I could always change this while at it.